File: examples/heat_example/heat_no_mpi.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/timeb.h>
#include <zmq.h>
#include <melissa_api_no_mpi.h>


void read_file(int*   ,
               int*   ,
               double*,
               double*,
               double* );

void init(double*,
          int*   ,
          double* );

void filling_A(double*,
               double*,
               double*,
               double*,
               double* );

void filling_F(int*   ,
               int*   ,
               double*,
               double*,
               double*,
               double*,
               double*,
               double*,
               double*,
               int*   ,
               double*,
               double*,
               double* );

void conjgrad(double*,
              double*,
              double*,
              int*   ,
              int*   ,
              double* );

void finalize(double*,
              double*,
              int*   ,
              int*   ,
              int*   ,
              double*,
              double* );


int main( int argc, char **argv )
{

  int    nx, ny, n, nmax, nb_op;
  double lx, ly, dt, dx, dy, d, t, epsilon, t1, t2, temp = 0;
  double *u = NULL;
  double *f = NULL;
  double a[3];
  double param[5];
  int sobol_rank = 0;
  int sobol_group = 0;
  char *field_name = "heat";
  struct timeb tp;

  for (n=0; n<5; n++)
      param[n] = 0;
    if (argc > n+1)
    {
       param[n] = strtod(argv[n+1], NULL);
    }
  if (argc > 1)
  {
      temp = param[0];
  }
  if (argc > 3)
  {
    sobol_rank  = (int)strtod(argv[argc-2], NULL);
    sobol_group = (int)strtod(argv[argc-1], NULL);
  }

  ftime(&tp);
  t1 = (double)tp.time + (double)tp.millitm / 1000;

  read_file(&nx,&ny,&lx,&ly,&d);

  nb_op   = nx*ny;
  dt      = 0.01;
  nmax    = 100;
  dx      = lx/(nx+1);
  dy      = ly/(ny+1);
  epsilon = 0.0001;

  u = malloc(nb_op * sizeof(double));
  f = malloc(nb_op * sizeof(double));
  init(&u[0], &nb_op, &temp);
  filling_A (&d, &dx, &dy, &dt, &a[0]); /* fill A */

  melissa_init_no_mpi(&nb_op, &sobol_rank, &sobol_group);

  for(n=1;n<=nmax;n++)
  {
    t+=dt;
    filling_F (&nx, &ny, &u[0], &d, &dx, &dy, &dt, &t, &f[0], &nb_op, &lx, &ly, &param[0]);
    conjgrad (&a[0], &f[0], &u[0], &nx, &ny, &epsilon);
    melissa_send_no_mpi(&n, field_name, u, &sobol_rank, &sobol_group);
  }

  finalize (&dx, &dy, &nx, &ny, &nb_op, &u[0], &f[0]);

  melissa_finalize ();

  ftime(&tp);
  t2 = (double)tp.time + (double)tp.millitm / 1000;

  fprintf(stdout, "Calcul time: %g sec\n", t2-t1);
  fprintf(stdout, "Final time step: %g\n", t);

  return 0;
}

File: examples/heat_example/heat_no_mpi.f90
 
program heat_no_mpi

  use heat_utils_no_mpi

  implicit none

  include "melissa_api.f90"

  integer :: nx, ny, n, nmax, nb_op, narg
  real*8 :: lx, ly, dt, dx, dy, d, t, epsilon, t1, t2, temp = 0
  real*8,dimension(:),pointer :: U => null(), F => null()
  real*8,dimension(3) :: A
  real*8,dimension(5) :: param
  character(len=32) :: arg
  character(len=5) :: name = C_CHAR_"heat"//C_NULL_CHAR
  integer :: sobol_rank = 0, sobol_group = 0

  narg = iargc()
  param(:) = 0
  do n=1, 5
    if(narg.ge.n) then
      call getarg(n, arg)
      read( arg, * ) param(n)
    endif
  enddo
  ! initial temperature
  temp = param(1)
  if(narg.ge.3) then
    call getarg(narg - 1, arg)
    read( arg, * ) sobol_rank ! sobol rank
    call getarg(narg, arg)
    read( arg, * ) sobol_group ! sobol group
  endif

  call cpu_time(t1)

  call read_file(nx, ny, lx, ly, d)

  nb_op   = nx*ny
  dt      = 0.01
  nmax    = 100
  dx      = lx/(nx+1)
  dy      = ly/(ny+1)
  epsilon = 0.0001

  allocate(U(nb_op))
  allocate(F(nb_op))
  call init(U, nb_op, temp)
  call filling_A(d, dx, dy, dt, A) ! fill A

  call melissa_init_no_mpi(nb_op, sobol_rank, sobol_group)

  do n=1, nmax
    t = t + dt
    call filling_F(nx, ny, U, d, dx, dy, dt, t, F, nb_op, lx, ly, param)
    call conjgrad(A, F, U, nx, ny, epsilon)

    call melissa_send_no_mpi(n, name, u, sobol_rank, sobol_group)
  end do

  call finalize(dx, dy, nx, ny, nb_op, u, f)

  call melissa_finalize()
  
  call cpu_time(t2)
  print*,'Calcul time:', t2-t1, 'sec'
  print*,'Final time step:', t
  
  deallocate(U, F)

end program heat_no_mpi